File: LocalPods/DialogView/DialogView/Core/Dialog+Common.m
//
//  Dialog+Common.m
//  TCLPlus
//
//  Created by OwenChen on 2021/1/4.
//  Copyright © 2021 TCL Electronics Holdings Ltd. All rights reserved.
//

#import "DialogModelEditor.h"

#import "Dialog+Common.h"


@implementation Dialog (Common)

/// 显示普通标题和副标题
+ (DialogView *)showDialogWithTitle:(NSString *)title
                           subtitle:(NSString *)subtitle
                            buttons:(NSArray<NSString *> *)buttons
                        resultBlock:(DialogResultBlock)resultBlock {
    DialogView *view = [self dialogWithTitle:title
                                    subtitle:subtitle
                                     buttons:buttons
                                 resultBlock:resultBlock];
    [self addDialogView:view];
    return view;
}

/// 显示普通标题和副标题（可定义按钮布局）
+ (DialogView *)showDialogWithTitle:(NSString *)title
                           subtitle:(NSString *)subtitle
                            buttons:(NSArray<NSString *> *)buttons
                  buttonsLayoutType:(DialogButtonsLayoutType)buttonsLayoutType
                        resultBlock:(DialogResultBlock)resultBlock {
    DialogView *view = [self dialogWithTitle:title
                                    subtitle:subtitle
                                     buttons:buttons
                           buttonsLayoutType:buttonsLayoutType
                                 resultBlock:resultBlock];
    [self addDialogView:view];
    return view;
}

/// 显示普通标题和副标题（副标题是富文本）
+ (DialogView *)showDialogWithTitle:(NSString *)title
                 attributedSubtitle:(NSAttributedString *)attributedSubtitle
                            buttons:(NSArray<NSString *> *)buttons
                        resultBlock:(DialogResultBlock)resultBlock {
    DialogView *view = [self dialogWithTitle:title
                          attributedSubtitle:attributedSubtitle
                                     buttons:buttons
                                 resultBlock:resultBlock];
    [self addDialogView:view];

    return view;
}

/// 普通标题和副标题
+ (DialogView *)dialogWithTitle:(NSString *)title
                       subtitle:(NSString *)subtitle
                        buttons:(NSArray<NSString *> *)buttons
                    resultBlock:(DialogResultBlock)resultBlock {
    return [self dialogWithTitle:title
                        subtitle:subtitle
                         buttons:buttons
               buttonsLayoutType:DialogButtonsLayoutHorizontal
                     resultBlock:resultBlock];
}

/// 普通标题和副标题（可定义按钮布局）
+ (DialogView *)dialogWithTitle:(NSString *)title
                       subtitle:(NSString *)subtitle
                        buttons:(NSArray<NSString *> *)buttons
              buttonsLayoutType:(DialogButtonsLayoutType)buttonsLayoutType
                    resultBlock:(DialogResultBlock)resultBlock {
    NSMutableArray *array = [NSMutableArray array];
    if (title.length > 0) {
        TextDialogModel *textModel = [DialogModelEditor createTextDialogModelWithTitle:title titleColor:nil];
        [array addObject:textModel];
    }

    if (subtitle.length > 0) {
        TextDialogModel *textModel1 = [DialogModelEditor createTextDialogModelWithSubtitle:subtitle subtitleColor:nil];
        [array addObject:textModel1];
    }

    [DialogModelEditor createModelMarginWithModels:array existTitle:title.length > 0];

    if (buttons.count > 0) {
        ButtonsDialogModel *buttonsModel = [DialogModelEditor createButtonsDialogModelWithButtons:buttons buttonColors:nil buttonsLayoutType:buttonsLayoutType];
        [array addObject:buttonsModel];
    }

    DialogView *view = [[DialogView alloc] initWithDialogModels:array
                                                      animation:[Dialog sharedInstance].animationType
                                                       position:DialogPositionMiddle
                                                        sideTap:YES
                                                         bounce:YES];
    view.resultBlock = resultBlock;

    return view;
}

/// 普通标题和副标题（副标题是富文本）
+ (DialogView *)dialogWithTitle:(NSString *)title
             attributedSubtitle:(NSAttributedString *)attributedSubtitle
                        buttons:(NSArray<NSString *> *)buttons
                    resultBlock:(DialogResultBlock)resultBlock {
    NSMutableArray *array = [NSMutableArray array];
    if (title.length > 0) {
        TextDialogModel *textModel = [DialogModelEditor createTextDialogModelWithTitle:title titleColor:nil];
        [array addObject:textModel];
    }

    if (attributedSubtitle.string.length > 0) {
        TextDialogModel *textModel1 = [DialogModelEditor createTextDialogModelWithAttributedSubtitle:attributedSubtitle subtitleColor:nil];
        [array addObject:textModel1];
    }

    [DialogModelEditor createModelMarginWithModels:array existTitle:title.length > 0];

    if (buttons.count > 0) {
        ButtonsDialogModel *buttonsModel = [DialogModelEditor createButtonsDialogModelWithButtons:buttons buttonColors:nil];
        [array addObject:buttonsModel];
    }

    DialogView *view = [[DialogView alloc] initWithDialogModels:array
                                                      animation:[Dialog sharedInstance].animationType
                                                       position:DialogPositionMiddle
                                                        sideTap:YES
                                                         bounce:YES];
    view.resultBlock = resultBlock;

    return view;
}

@end

File: LocalPods/DialogViewSwift/DialogViewSwift/Core/Dialog+Common.swift
//
//  Dialog+Common.swift
//  TCLHome
//
//  Created by huaizhang.chen on 2022/2/23.
//  Copyright © 2022 TCL Eagle Lab. All rights reserved.
//

import Foundation

// MARK: 自动显示弹框

extension EagleLabKit where Base: Dialog {
    /// 显示普通标题和副标题
    /// - Parameters:
    ///   - title: 主标题
    ///   - subtitle: 副标题
    ///   - buttons: 按钮标题
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    @discardableResult
    static func showDialog(title: String,
                           subtitle: String,
                           buttons: [String]?,
                           resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title:
            title,
            subtitle:
            subtitle,
            buttons: buttons,
            resultBlock: resultBlock)
        Dialog.addDialog(dialogView: view)
        return view
    }

    ///  显示普通标题和副标题（可定义按钮布局）
    /// - Parameters:
    ///   - title: 主标题
    ///   - subtitle: 副标题
    ///   - buttons: 按钮标题
    ///   - buttonsLayoutType: 按钮布局样式
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    @discardableResult
    static func showDialog(title: String,
                           subtitle: String,
                           buttons: [String]?,
                           buttonsLayoutType: DialogButtonsLayoutType,
                           resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title: title,
                          subtitle: subtitle,
                          buttons: buttons,
                          buttonsLayoutType: buttonsLayoutType,
                          resultBlock: resultBlock)
        Dialog.addDialog(dialogView: view)
        return view
    }

    /// 显示普通标题和副标题（副标题是富文本）
    /// - Parameters:
    ///   - title: 主标题
    ///   - attributedSubtitle: 富文本副标题
    ///   - buttons: 按钮标题
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    @discardableResult
    static func showDialog(title: String,
                           attributedSubtitle: NSAttributedString?,
                           buttons: [String]?,
                           resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title: title,
                          attributedSubtitle: attributedSubtitle,
                          buttons: buttons,
                          resultBlock: resultBlock)
        Dialog.addDialog(dialogView: view)
        return view
    }
}

// MARK: 生成弹框，不自动显示

extension EagleLabKit where Base: Dialog {
    /// 普通标题和副标题
    /// - Parameters:
    ///   - title: 主标题
    ///   - subtitle: 副标题
    ///   - buttons: 按钮标题
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    static func dialog(title: String,
                       subtitle: String,
                       buttons: [String]?,
                       resultBlock: DialogView.DialogViewResult?) -> DialogView {
        let view = dialog(title: title,
                          subtitle: subtitle,
                          buttons: buttons,
                          buttonsLayoutType: .horizontal,
                          resultBlock: resultBlock)
        return view
    }

    ///  普通标题和副标题（可定义按钮布局）
    /// - Parameters:
    ///   - title: 主标题
    ///   - subtitle: 副标题
    ///   - buttons: 按钮标题
    ///   - buttonsLayoutType: 按钮布局样式
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    static func dialog(title: String,
                       subtitle: String,
                       buttons: [String]?,
                       buttonsLayoutType: DialogButtonsLayoutType,
                       resultBlock: DialogView.DialogViewResult?) -> DialogView {
        var array: [BasicDialogModel] = []

        if !title.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(title: title, titleColor: "")
            array.append(textModel)
        }

        if !subtitle.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(subtitle: subtitle, subtitleColor: "")
            array.append(textModel)
        }

        DialogModelEditor.createModelMargin(models: &array, existTitle: !title.isEmpty)

        if let buttonsArray = buttons, !buttonsArray.isEmpty {
            let buttonsModel = DialogModelEditor.createButtonsDialogModel(buttons: buttonsArray,
                                                                          buttonsLayoutType: buttonsLayoutType)
            array.append(buttonsModel)
        }

        let view = DialogView()
        view.configDialogView(dialogModels: array)
        view.dialogResult = resultBlock
        return view
    }

    /// 普通标题和副标题（副标题是富文本）
    /// - Parameters:
    ///   - title: 主标题
    ///   - attributedSubtitle: 富文本副标题
    ///   - buttons: 按钮标题
    ///   - resultBlock: 结果回调
    /// - Returns: DialogView
    static func dialog(title: String,
                       attributedSubtitle: NSAttributedString?,
                       buttons: [String]?,
                       resultBlock: DialogView.DialogViewResult?) -> DialogView {
        var array: [BasicDialogModel] = []

        if !title.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(title: title, titleColor: "")
            array.append(textModel)
        }

        if let attributedSubText = attributedSubtitle, !attributedSubText.string.isEmpty {
            let textModel = DialogModelEditor.createTextDialogModel(attributedSubtitle: attributedSubText, subtitleColor: "")
            array.append(textModel)
        }

        DialogModelEditor.createModelMargin(models: &array, existTitle: !title.isEmpty)

        if let buttonsArray = buttons, !buttonsArray.isEmpty {
            let buttonsModel = DialogModelEditor.createButtonsDialogModel(buttons: buttonsArray)
            array.append(buttonsModel)
        }

        let view = DialogView()
        view.configDialogView(dialogModels: array)
        view.dialogResult = resultBlock
        return view
    }
}
